lib/fastlane/plugin/wpmreleasetoolkit/actions/common/buildkite_upload_pipeline_action.rb

```diff
@@ -0,0 +1,70 @@
+module Fastlane
+  module Actions
+    class BuildkiteUploadPipelineAction < Action
+      def self.run(params)
+        env_file = params[:env_file]
+        pipeline_file = params[:pipeline_file]
+        branch = params[:branch]
+        commit = params[:commit]
+
+        UI.user_error!("Pipeline file not found: #{pipeline_file}") unless File.exist?(pipeline_file)
+
+        UI.message "Uploading pipeline #{pipeline_file} on branch #{branch}, commit #{commit}"
+
+        ENV['BUILDKITE_BRANCH'] = branch
+        ENV['BUILDKITE_COMMIT'] = commit
+
+        if env_file && File.exist?(env_file)
+          UI.message("Sourcing environment file: #{env_file}")
+
+          sh(". #{env_file} && buildkite-agent pipeline upload #{pipeline_file}")
+        else
+          sh('buildkite-agent', 'pipeline', 'upload', pipeline_file)
+        end
+      end
+
+      def self.description
+        # https://buildkite.com/docs/agent/v3/cli-pipeline#uploading-pipelines
+        'Uploads a pipeline to Buildkite, adding all its steps to the current build'
+      end
+
+      def self.available_options
+        [
+          FastlaneCore::ConfigItem.new(
+            key: :env_file,
+            description: 'The path to the environment variables file to be sourced before uploading the pipeline',
+            optional: true,
+            type: String
+          ),
+          FastlaneCore::ConfigItem.new(
+            key: :pipeline_file,
+            description: 'The path to the Buildkite pipeline file',
+            optional: false,
+            type: String
+          ),
+          FastlaneCore::ConfigItem.new(
+            key: :branch,
+            description: 'The branch you want to run the pipeline on',
+            optional: false,
```

Comment:
At first when I started reading the code before I reached this point, I thought that `branch` would be optional (and in such case, the steps added by `buildkite-agent pipeline upload` would just run on the same branch as the build on which that action was called). This also explains my questions on L14–L15 above.

While it's likely that we'll want to provide the `branch` more often than not in the main use cases we'll currently have for this action, I think there are also valid use cases for which we'd just want to upload additional steps to the current pipeline on the same branch.

There are actually already [use cases for that (= calling `pipeline upload` from steps but still on the same branch) in Beeper Desktop](https://github.com/beeper/beeper-desktop-new/blob/2af519081993f9552ec429f4c532581a72f55917/.buildkite/pipelines/official-build.yml#L44-L53) for example. _Sure for now they call `buildkite-agent pipeline upload` directly from the `command:`, but if one day they wanted to migrate that logic into a dedicated _fastlane_ lane and start using this action for that…_

Reply:
True, good point. Updated here: 7349eb36526b2d7e1d0b95beb822d0399ab4a6f4